{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6bcb5235-f307-48bf-87b0-e9e0e56a8f27",
   "metadata": {},
   "source": [
    "# decision tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "7ed8282c-9a58-4401-8973-05d69f415773",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from Bio import *\n",
    "import matplotlib.pyplot as plt\n",
    "import warnings\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "from sklearn import tree\n",
    "from sklearn.tree import plot_tree\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import confusion_matrix\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82a3a1c8-c9fe-4705-aed5-c44a4a5db2b6",
   "metadata": {},
   "source": [
    "here, we are required to merge the bigbigdf from the snippy analysis to the phenotype of the phages on 1 defence system. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f6213ba6-4104-4f79-8c01-43b07e05e479",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def find_key_gene_db_name(filtered_intergenic, startstopdict, position):\n",
    "    found = 0\n",
    "    for key in startstopdict.keys(): \n",
    "        array = startstopdict[key]\n",
    "        if position in array: \n",
    "            found = 1\n",
    "            protein_annotation = filtered_intergenic.loc[key, 'protein annotation']\n",
    "            db_name = filtered_intergenic.loc[key, 'db name']\n",
    "    if found == 0: \n",
    "        protein_annotation = 'no gene'\n",
    "        db_name = 'no gene'\n",
    "    return protein_annotation, db_name\n",
    "\n",
    "def filter_snippy_df(morphology_reference, morphology_reads, reference_phage, inter_or_intra): \n",
    "    myo_snippy_df = snippy_df[(snippy_df.loc[:, 'morphology reference'] == morphology_reference) & (snippy_df.loc[:, 'morphology read'] == morphology_reads)]\n",
    "    pa10myo_snippy_df = myo_snippy_df[myo_snippy_df.loc[:, 'reference'] == reference_phage]\n",
    "    pa10myo_snippy_df = pa10myo_snippy_df[pa10myo_snippy_df.loc[:, 'inter or intra'] == inter_or_intra]\n",
    "    return pa10myo_snippy_df\n",
    "\n",
    "def filter_intergenic_df(phage): \n",
    "    df = gbk_df[gbk_df.loc[:, 'phage'] == phage]\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "eba152c3-9b3c-461b-9144-3ec7ed91f742",
   "metadata": {},
   "outputs": [],
   "source": [
    "snippy_df = pd.read_csv(r'C:\\Users\\Halewijn\\OneDrive\\Documenten\\third year\\BEP\\thesis\\mutational analysis\\final_snippy_analysis'\n",
    "           + '/' + 'snippy_df.csv')\n",
    "gbk_df = pd.read_csv(r'C:\\Users\\Halewijn\\OneDrive\\Documenten\\third year\\BEP\\thesis\\mutational analysis\\compressed dataframes' + '/' + \n",
    "                    \"intergenic_information_df.csv\")\n",
    "phenotype_df = pd.read_csv(r'C:\\Users\\Halewijn\\OneDrive\\Documenten\\third year\\BEP\\thesis\\mutational analysis\\make phenotypic array' + \n",
    "                          '/' + 'binary_scaled_myo_plate_reader_df.csv').set_index('phage')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "75310c16-5303-45ba-8f36-920a2fa09c04",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dictionary = {}\n",
    "# start_stop_big_array = []\n",
    "# for ii in range(gbk_df.shape[0]): \n",
    "#     start = gbk_df.loc[ii, 'start']\n",
    "#     stop = gbk_df.loc[ii, 'stop']\n",
    "#     dictionary[ii] = np.arange(start, stop + 1, 1)\n",
    "    \n",
    "# gbk_df.loc[:, 'start stop pos array'] = dictionary.values()\n",
    "\n",
    "# #from this point on, we filter further, on the myos and specifically Pa10. \n",
    "# phage = 'Pa12'\n",
    "# filtered_snippy = filter_snippy_df('Myoviridae', 'Myoviridae', phage, 'inter').reset_index(drop = True)\n",
    "# filtered_intergenic = filter_intergenic_df(phage = phage)\n",
    "# startstopdict = dict(filtered_intergenic.loc[:, 'start stop pos array'])\n",
    "\n",
    "# #here, I merge the dataset with the gene the snp finds itself in\n",
    "# for ii in range(filtered_snippy.shape[0]): \n",
    "#     position = filtered_snippy.loc[ii, 'POS']\n",
    "#     protein_annotation, db_name = find_key_gene_db_name(filtered_intergenic, startstopdict, position)\n",
    "#     filtered_snippy.loc[ii, 'protein annotation'] = protein_annotation\n",
    "#     filtered_snippy.loc[ii, 'db name'] = db_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "c8d297fe-0fb1-4c55-a1ea-35d24fba5cbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_list_snippy_files = os.listdir(r'C:\\Users\\Halewijn\\OneDrive\\Documenten\\third year\\BEP\\thesis\\mutational analysis\\compressed dataframes\\filtered_snippy_files')\n",
    "all_filtered_snippy = []\n",
    "for filtered_snippy_file in file_list_snippy_files: \n",
    "    filtered_snippy = pd.read_csv(r'C:\\Users\\Halewijn\\OneDrive\\Documenten\\third year\\BEP\\thesis\\mutational analysis\\compressed dataframes\\filtered_snippy_files' \n",
    "                                 + '/' + filtered_snippy_file)\n",
    "    filtered_snippy.loc[:, 'phage'] = [filtered_snippy_file[:-4]]*filtered_snippy.shape[0]\n",
    "    all_filtered_snippy = all_filtered_snippy + [filtered_snippy]\n",
    "all_snippy = pd.concat(all_filtered_snippy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "dfe84890-21cc-4c95-ac75-7a57e2f448c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAl</th>\n",
       "      <th>family</th>\n",
       "      <th>inter or intra</th>\n",
       "      <th>reference</th>\n",
       "      <th>reads</th>\n",
       "      <th>phage_x</th>\n",
       "      <th>morphology reference</th>\n",
       "      <th>phage_y</th>\n",
       "      <th>morphology read</th>\n",
       "      <th>protein annotation</th>\n",
       "      <th>db name</th>\n",
       "      <th>phage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>40848+(circular)</td>\n",
       "      <td>226</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>390.256</td>\n",
       "      <td>0</td>\n",
       "      <td>inter</td>\n",
       "      <td>Pa10</td>\n",
       "      <td>Pa12</td>\n",
       "      <td>Pa10</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Pa12</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>no gene</td>\n",
       "      <td>no gene</td>\n",
       "      <td>Pa10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>40848+(circular)</td>\n",
       "      <td>287</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1004.770</td>\n",
       "      <td>0</td>\n",
       "      <td>inter</td>\n",
       "      <td>Pa10</td>\n",
       "      <td>Pa12</td>\n",
       "      <td>Pa10</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Pa12</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Phage protein</td>\n",
       "      <td>['SEED:fig|287.19287.peg.2']</td>\n",
       "      <td>Pa10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>40848+(circular)</td>\n",
       "      <td>305</td>\n",
       "      <td>.</td>\n",
       "      <td>TGAC</td>\n",
       "      <td>CGAT</td>\n",
       "      <td>1175.090</td>\n",
       "      <td>0</td>\n",
       "      <td>inter</td>\n",
       "      <td>Pa10</td>\n",
       "      <td>Pa12</td>\n",
       "      <td>Pa10</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Pa12</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Phage protein</td>\n",
       "      <td>['SEED:fig|287.19287.peg.2']</td>\n",
       "      <td>Pa10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>40848+(circular)</td>\n",
       "      <td>331</td>\n",
       "      <td>.</td>\n",
       "      <td>ACA</td>\n",
       "      <td>CTG</td>\n",
       "      <td>1144.870</td>\n",
       "      <td>0</td>\n",
       "      <td>inter</td>\n",
       "      <td>Pa10</td>\n",
       "      <td>Pa12</td>\n",
       "      <td>Pa10</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Pa12</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Phage protein</td>\n",
       "      <td>['SEED:fig|287.19287.peg.2']</td>\n",
       "      <td>Pa10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>40848+(circular)</td>\n",
       "      <td>350</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>1481.740</td>\n",
       "      <td>0</td>\n",
       "      <td>inter</td>\n",
       "      <td>Pa10</td>\n",
       "      <td>Pa12</td>\n",
       "      <td>Pa10</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Pa12</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Phage protein</td>\n",
       "      <td>['SEED:fig|287.19287.peg.2']</td>\n",
       "      <td>Pa10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10612</th>\n",
       "      <td>10612</td>\n",
       "      <td>61662+(circular)</td>\n",
       "      <td>65486</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1639.000</td>\n",
       "      <td>0</td>\n",
       "      <td>inter</td>\n",
       "      <td>Pa49</td>\n",
       "      <td>Pa10</td>\n",
       "      <td>Pa49</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Pa10</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Phage protein</td>\n",
       "      <td>['SEED:fig|287.19304.peg.92']</td>\n",
       "      <td>Pa49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10613</th>\n",
       "      <td>10613</td>\n",
       "      <td>61662+(circular)</td>\n",
       "      <td>65495</td>\n",
       "      <td>.</td>\n",
       "      <td>GGCG</td>\n",
       "      <td>AGCA</td>\n",
       "      <td>1586.710</td>\n",
       "      <td>0</td>\n",
       "      <td>inter</td>\n",
       "      <td>Pa49</td>\n",
       "      <td>Pa10</td>\n",
       "      <td>Pa49</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Pa10</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Phage protein</td>\n",
       "      <td>['SEED:fig|287.19304.peg.92']</td>\n",
       "      <td>Pa49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10614</th>\n",
       "      <td>10614</td>\n",
       "      <td>61662+(circular)</td>\n",
       "      <td>65554</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1586.190</td>\n",
       "      <td>0</td>\n",
       "      <td>inter</td>\n",
       "      <td>Pa49</td>\n",
       "      <td>Pa10</td>\n",
       "      <td>Pa49</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Pa10</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Phage protein</td>\n",
       "      <td>['SEED:fig|287.19304.peg.92']</td>\n",
       "      <td>Pa49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10615</th>\n",
       "      <td>10615</td>\n",
       "      <td>61662+(circular)</td>\n",
       "      <td>65591</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>1712.520</td>\n",
       "      <td>0</td>\n",
       "      <td>inter</td>\n",
       "      <td>Pa49</td>\n",
       "      <td>Pa10</td>\n",
       "      <td>Pa49</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Pa10</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Phage protein</td>\n",
       "      <td>['SEED:fig|287.19304.peg.92']</td>\n",
       "      <td>Pa49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10616</th>\n",
       "      <td>10616</td>\n",
       "      <td>61662+(circular)</td>\n",
       "      <td>65819</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>518.053</td>\n",
       "      <td>0</td>\n",
       "      <td>inter</td>\n",
       "      <td>Pa49</td>\n",
       "      <td>Pa10</td>\n",
       "      <td>Pa49</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Pa10</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>no gene</td>\n",
       "      <td>no gene</td>\n",
       "      <td>Pa49</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>123487 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0             CHROM    POS ID   REF   ALT      QUAl family  \\\n",
       "0              0  40848+(circular)    226  .     C     T   390.256      0   \n",
       "1              1  40848+(circular)    287  .     C     A  1004.770      0   \n",
       "2              2  40848+(circular)    305  .  TGAC  CGAT  1175.090      0   \n",
       "3              3  40848+(circular)    331  .   ACA   CTG  1144.870      0   \n",
       "4              4  40848+(circular)    350  .     C     G  1481.740      0   \n",
       "...          ...               ...    ... ..   ...   ...       ...    ...   \n",
       "10612      10612  61662+(circular)  65486  .     G     A  1639.000      0   \n",
       "10613      10613  61662+(circular)  65495  .  GGCG  AGCA  1586.710      0   \n",
       "10614      10614  61662+(circular)  65554  .     T     C  1586.190      0   \n",
       "10615      10615  61662+(circular)  65591  .     T     G  1712.520      0   \n",
       "10616      10616  61662+(circular)  65819  .     G     A   518.053      0   \n",
       "\n",
       "      inter or intra reference reads phage_x morphology reference phage_y  \\\n",
       "0              inter      Pa10  Pa12    Pa10           Myoviridae    Pa12   \n",
       "1              inter      Pa10  Pa12    Pa10           Myoviridae    Pa12   \n",
       "2              inter      Pa10  Pa12    Pa10           Myoviridae    Pa12   \n",
       "3              inter      Pa10  Pa12    Pa10           Myoviridae    Pa12   \n",
       "4              inter      Pa10  Pa12    Pa10           Myoviridae    Pa12   \n",
       "...              ...       ...   ...     ...                  ...     ...   \n",
       "10612          inter      Pa49  Pa10    Pa49           Myoviridae    Pa10   \n",
       "10613          inter      Pa49  Pa10    Pa49           Myoviridae    Pa10   \n",
       "10614          inter      Pa49  Pa10    Pa49           Myoviridae    Pa10   \n",
       "10615          inter      Pa49  Pa10    Pa49           Myoviridae    Pa10   \n",
       "10616          inter      Pa49  Pa10    Pa49           Myoviridae    Pa10   \n",
       "\n",
       "      morphology read protein annotation                        db name phage  \n",
       "0          Myoviridae            no gene                        no gene  Pa10  \n",
       "1          Myoviridae      Phage protein   ['SEED:fig|287.19287.peg.2']  Pa10  \n",
       "2          Myoviridae      Phage protein   ['SEED:fig|287.19287.peg.2']  Pa10  \n",
       "3          Myoviridae      Phage protein   ['SEED:fig|287.19287.peg.2']  Pa10  \n",
       "4          Myoviridae      Phage protein   ['SEED:fig|287.19287.peg.2']  Pa10  \n",
       "...               ...                ...                            ...   ...  \n",
       "10612      Myoviridae      Phage protein  ['SEED:fig|287.19304.peg.92']  Pa49  \n",
       "10613      Myoviridae      Phage protein  ['SEED:fig|287.19304.peg.92']  Pa49  \n",
       "10614      Myoviridae      Phage protein  ['SEED:fig|287.19304.peg.92']  Pa49  \n",
       "10615      Myoviridae      Phage protein  ['SEED:fig|287.19304.peg.92']  Pa49  \n",
       "10616      Myoviridae            no gene                        no gene  Pa49  \n",
       "\n",
       "[123487 rows x 18 columns]"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_snippy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "032bcc61-0955-44d6-b1ed-40165d1d4e15",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Halewijn\\AppData\\Local\\Temp/ipykernel_19148/2794071930.py:2: FutureWarning: Passing 'suffixes' which cause duplicate columns {'phage_x'} in the result is deprecated and will raise a MergeError in a future version.\n",
      "  ds_ref_all_snippy = pd.merge(ref_all_snippy, phenotype_df.loc[:, 'Zorya I'].reset_index(), left_on='reads', right_on='phage')\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'reads'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\pandas\\core\\indexes\\base.py\u001b[0m in \u001b[0;36mget_loc\u001b[1;34m(self, key, method, tolerance)\u001b[0m\n\u001b[0;32m   3360\u001b[0m             \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 3361\u001b[1;33m                 \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   3362\u001b[0m             \u001b[1;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\pandas\\_libs\\index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\pandas\\_libs\\index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32mpandas\\_libs\\hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32mpandas\\_libs\\hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: 'reads'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_19148/2794071930.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[0mx_df\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mds_ref_all_snippy\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'POS'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'inter or intra'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'protein annotation'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'db name'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[0mdummy_x_df\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_dummies\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mx_df\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 5\u001b[1;33m \u001b[0mdummy_x_df\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'reads'\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mds_phenotype_df\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'reads'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      6\u001b[0m \u001b[1;31m#ds_ref_all_snippy.loc[:, 'reads']\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\pandas\\core\\indexing.py\u001b[0m in \u001b[0;36m__getitem__\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m    923\u001b[0m                 \u001b[1;32mwith\u001b[0m \u001b[0msuppress\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mKeyError\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mIndexError\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    924\u001b[0m                     \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_get_value\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m*\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mtakeable\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_takeable\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 925\u001b[1;33m             \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_getitem_tuple\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    926\u001b[0m         \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    927\u001b[0m             \u001b[1;31m# we by definition only have the 0th axis\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\pandas\\core\\indexing.py\u001b[0m in \u001b[0;36m_getitem_tuple\u001b[1;34m(self, tup)\u001b[0m\n\u001b[0;32m   1098\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_getitem_tuple\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mtup\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mtuple\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1099\u001b[0m         \u001b[1;32mwith\u001b[0m \u001b[0msuppress\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mIndexingError\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1100\u001b[1;33m             \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_getitem_lowerdim\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtup\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1101\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1102\u001b[0m         \u001b[1;31m# no multi-index, so validate all of the indexers\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\pandas\\core\\indexing.py\u001b[0m in \u001b[0;36m_getitem_lowerdim\u001b[1;34m(self, tup)\u001b[0m\n\u001b[0;32m    836\u001b[0m                 \u001b[1;31m# We don't need to check for tuples here because those are\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    837\u001b[0m                 \u001b[1;31m#  caught by the _is_nested_tuple_indexer check above.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 838\u001b[1;33m                 \u001b[0msection\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_getitem_axis\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    839\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    840\u001b[0m                 \u001b[1;31m# We should never have a scalar section here, because\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\pandas\\core\\indexing.py\u001b[0m in \u001b[0;36m_getitem_axis\u001b[1;34m(self, key, axis)\u001b[0m\n\u001b[0;32m   1162\u001b[0m         \u001b[1;31m# fall thru to straight lookup\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1163\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_validate_key\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1164\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_get_label\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1165\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1166\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_get_slice_axis\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mslice_obj\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mslice\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mint\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\pandas\\core\\indexing.py\u001b[0m in \u001b[0;36m_get_label\u001b[1;34m(self, label, axis)\u001b[0m\n\u001b[0;32m   1111\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_get_label\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mlabel\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mint\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1112\u001b[0m         \u001b[1;31m# GH#5667 this will fail if the label is not present in the axis.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1113\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mxs\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlabel\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1114\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1115\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_handle_lowerdim_multi_index_axis0\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mtup\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mtuple\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\pandas\\core\\generic.py\u001b[0m in \u001b[0;36mxs\u001b[1;34m(self, key, axis, level, drop_level)\u001b[0m\n\u001b[0;32m   3759\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0maxis\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;36m1\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3760\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mdrop_level\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 3761\u001b[1;33m                 \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   3762\u001b[0m             \u001b[0mindex\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3763\u001b[0m         \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\pandas\\core\\frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   3456\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mnlevels\u001b[0m \u001b[1;33m>\u001b[0m \u001b[1;36m1\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3457\u001b[0m                 \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 3458\u001b[1;33m             \u001b[0mindexer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   3459\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mis_integer\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3460\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[0mindexer\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\pandas\\core\\indexes\\base.py\u001b[0m in \u001b[0;36mget_loc\u001b[1;34m(self, key, method, tolerance)\u001b[0m\n\u001b[0;32m   3361\u001b[0m                 \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3362\u001b[0m             \u001b[1;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 3363\u001b[1;33m                 \u001b[1;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[0merr\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   3364\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3365\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mis_scalar\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mand\u001b[0m \u001b[0misna\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mand\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mhasnans\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: 'reads'"
     ]
    }
   ],
   "source": [
    "ref_all_snippy = all_snippy[all_snippy.loc[:, 'reference'] == 'Pa10']\n",
    "ds_ref_all_snippy = pd.merge(ref_all_snippy, phenotype_df.loc[:, 'Zorya I'].reset_index(), left_on='reads', right_on='phage')\n",
    "x_df = ds_ref_all_snippy[['POS', 'inter or intra', 'protein annotation', 'db name']]\n",
    "dummy_x_df = pd.get_dummies(x_df)\n",
    "dummy_x_df.loc[:, 'reads'] = list(ds_phenotype_df.loc[:, 'reads'])\n",
    "#ds_ref_all_snippy.loc[:, 'reads']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
